File: Lab3/Task2.m
B = 25;
A = 0.01;
fs = 5*B;

t_start=-10;
t_end= 10;
t = t_start:1/fs:t_end;

h_t = 2*0.1*sinc(2*B*t);
N = length(h_t);
h_f = fft(h_t, N)/fs;
h_f_abs = abs(h_f);
h_f = fftshift(h_f_abs);
freq_axis = linspace(-fs/2, fs/2, N);

m_t = sin(10*pi*t) + sin(40*pi*t);
m_f = fftshift(abs(fft(m_t)))/fs;

x_t = conv(h_t, m_t, 'same');
x_f = fftshift(abs(fft(x_t)))/fs;

figure;
subplot(5, 1, 1);
% title('Sinc Channel Effect');
plot(t, m_t);
grid on;
xlabel('Time(s)', 'interpreter', 'latex');
ylabel('Amplitude', 'interpreter', 'latex');
legend('$m(t)$', 'interpreter', 'latex');

subplot(5, 1, 2);
plot(t, x_t);
grid on;
xlabel('Time(s)', 'interpreter', 'latex');
ylabel('Amplitude', 'interpreter', 'latex');
legend('$x(t)$', 'interpreter', 'latex');

subplot(5, 1, 3);
plot(freq_axis, m_f);
grid on;
xlabel('Frequency(Hz)', 'interpreter', 'latex');
ylabel('Amplitude', 'interpreter', 'latex');
legend('$|M(f)|$', 'interpreter', 'latex');

subplot(5, 1, 4);
plot(freq_axis, x_f);
grid on;
xlabel('Frequency(Hz)', 'interpreter', 'latex');
ylabel('Amplitude', 'interpreter', 'latex');
legend('$|X(f)|$', 'interpreter', 'latex');

subplot(5, 1, 5);
plot(freq_axis, h_f);
grid on;
xlabel('Frequency(Hz)', 'interpreter', 'latex');
ylabel('Amplitude', 'interpreter', 'latex');
legend('$|H(f)|$', 'interpreter', 'latex');

File: Lab4/Task2.py
import numpy as np
import matplotlib.pyplot as plt
from numpy.fft import fft
import random
import time
from scipy.io import wavfile

start_time = 0
stop_time = 1
N = 8  # Maximum frequency component in Hertz for the given spectrum
fs = 10 * N
ts = 1 / fs
mx = np.array([])
tvect = np.array([])
PI = np.pi
T_total = 30

(afs, audio) = wavfile.read('sample_a.wav')
aud_sample = audio[:, 0]

aud_sample = 0.1*aud_sample[0:T_total*fs]

def m1(t):
    return np.cos(2*PI*N*t)

def m2(t):
    return 2*N*np.sinc(2*N*t)

def m3(t):
    return (np.abs(t)<1).astype(float)

def m4(t):
    start = int(t[0]*fs)
    stop = start + t.size
    return aud_sample[start:stop]

msgs = [m1, m2, m3, m4]



for T in range(T_total):
    time = np.arange(start_time+T, stop_time+T, 1/fs)
    func = np.random.choice(msgs)
    m_t = func(time)
    mx = np.append(mx, m_t)
    tvect = np.append(tvect, time)
    mf = fft(mx)/fs
    N = len(mf)
    mf_abs_sorted = np.fft.fftshift(abs(mf))
    freq_axis = np.linspace(-fs/2, fs/2, N)

    # Time Domain
    plt.figure(1)
    plt.plot(tvect, mx)
    plt.title(f'Time Domain, Time=({start_time+T}, {stop_time+T})')
    plt.xlabel('time')
    plt.ylabel('Amplitude')

    # Save the final figure
    if T==T_total-1:
        plt.savefig('task2_final_time.svg')

    # Frequency Domain
    plt.figure(2)
    plt.plot(freq_axis, mf_abs_sorted)
    plt.xlabel('Frequency(Hz)')
    plt.ylabel('Amplitude')
    plt.title(f'Frequency Domain, Time=({start_time+T}, {stop_time+T})')

    # Save the final figure
    if T==T_total-1:
        plt.savefig('task2_final_fft.svg')

    plt.show()
    plt.pause(0.5)